### PR TITLE
Vanilla recurrent neural network layer implementation

### PR DESCRIPTION
**Pull request by cmendl**
This feature implements a vanilla recurrent neural network layer (analogous to the convolutional layer). The input is treated as a sequence of values, and the output consists of the corresponding sequence of hidden states, which may be further processed by upstream layers. Vanilla RNNs typically use tanh as nonlinearity; the implementation also supports general activation functions via activations.hpp.

### CODE REVIEW

**Comment by gcarleo** [comment]
Thanks a lot @cmendl , this looks nice already. The tests seem to fail because of precision issues, see line below. Is it expected to have these awfully large numbers there for the derivatives? @kchoo1118 does this happen also for other layers? 

`
:   REQUIRE( Approx(std::exp(std::imag(numder))).epsilon(eps * 1000) == std::exp(std::imag(ders(p))) )

11: with expansion:

11:   Approx(

11:   2481611622392030284204041021520096040991091520412639068486229053347482124885-

11:   7069221178874666754475862373155316121389901187996713846789285644750107689740-

11:   6884829711368886058646321877566625118377528409078439702673459912755910392596-

11:   66765966213120.0 )

11:   ==

11:   2500828599482244218506420625306335476258406427216016958548222570082355546877-

11:   2563555369749358154005564047634075738210400490182726794023951750185667222740-

11:   9152312355992620449363478567584658259566014682950608204641563226908558134210-

11:   21024445530112.0
`

**Comment by cmendl** [comment]
The large numbers are likely due to the nested evaluation of the nonlinearity in the RNN layer. This could be related to the unstable gradient problem of deep neural networks... Maybe one could alleviate the issue using a finite "window" for backpropagation...?

**Comment by kchoo1118** [comment]
Yes mostly the initialization (sigma =1) of the parameters is too large in the tests

**Comment by gcarleo** [comment]
It's maybe then a good idea to put a smaller sigma for the tests...

File: Tutorials/AKLT/AKLT.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# limitations under the License.
+
+from __future__ import print_function
+import json
+import numpy as np
+
+sigmax = [[0, 1, 0], [1, 0, 1], [0, 1, 0]]
+sigmay = [[0, -1j, 0], [1j, 0, -1j], [0, 1j, 0]]
+sigmaz = [[1, 0, 0], [0,0,0], [0, 0, -1]]
+
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply Tutorials/AKLT/AKLT.py]: Important detail I forgot: 
```python
pars['Graph'] = {
    'Name': 'Hypercube',
    'L': L,
    'Dimension': 1,
    'Pbc': True,
}

pars['Hilbert'] = {
    'S': 1,
    'Nspins': L,
    'TotalSz':0,
    'Name': "Spin",
}

# defining a custom bond hamiltonian
pars['Hamiltonian'] = {
    'Name': 'Graph',
    'BondOps': [(heisenberg).tolist(),(heisenberg_squared/3.).tolist()],
}
```

[the remark this answers is not in the document]
Comment (everthemore) [reply Tutorials/AKLT/AKLT.py]: Looks much nicer indeed, thanks.  We've managed to train an RNN on the AKLT model before (which is why I included this as a benchmark in the first place). Will try to check the gradients and see what's going on.

[the remark this answers is not in the document]
Comment (kchoo1118) [reply Tutorials/AKLT/AKLT.py]: Maybe try smaller initial parameters?

**Comment by cmendl** [comment]
To improve the capability of the RNN layer, we could maybe encode the input at each site as unit vectors (with length equal to the local Hilbert space dimension). For spin 1/2 for example, currently the input at each site is simply 1 or -1 (corresponding to spin up or down) (right?), which would become (1,0) or (0,1). In terms of the RNN layer, the column dimension of the U matrix would then be equal to the local Hilbert space dimension (instead of currently 1). I guess that's similar to rbm_spin.hpp versus rbm_multival.hpp... (BTW is the InputChannels parameter in conv_layer.hpp intended as local Hilbert space dimension?) In practical terms, one could maybe insert a simple layer between the physical input and the RNN layer which performs the encoding as unit vectors...

File: Test/Machine/layer_input_tests.hpp [hunk 2]
```diff
@@ -40,6 +40,18 @@ std::vector<netket::json> GetLayerInputs() {
           {"Hamiltonian", {{"Name", "Heisenberg"}}}};
   input_tests.push_back(pars);
 
+  // Recurrent layer
+  pars = {{"Graph",
```
Comment (gcarleo) [inline Test/Machine/layer_input_tests.hpp]: @kchoo1118 where is the sigma of the random parameters set in the testing? You mentioned it is sigma=1 by default?

**Comment by gcarleo** [comment]
@cmendl , that might be an improvement to consider, yes. @everthemore mentioned they managed to learn the AKLT before using a RNN. Was it with binary input variables? 
Anyway, I think there is still some other (major) optimization issue, since even for the Ising model the energy just stays stuck at the initial value. I think it is just that the initial parameters are too large/too small, and the gradient doesn't do much.

File: Test/Machine/layer_input_tests.hpp (hunk 2 above)
Comment (kchoo1118) [reply Test/Machine/layer_input_tests.hpp]: line 109 in unit-machine.cc

**Comment by cmendl** [comment]
@gcarleo Yes @everthemore and myself did a small toy model simulation for the AKLT model with 3D input variables (due to spin 1). But actually the layer we were able to "train" successfully was a "matrix product state layer", i.e., using a 3D weight tensor (also compare with Eq. (2.3) in https://arxiv.org/abs/1710.09431). This is not too surprising since the AKLT state can be written exactly as MPS. By the way such a MPS layer could be a viable alternative or replacement for the vanilla RNN layer..?

**Comment by cmendl** [comment]
The last commit introduces a "local Hilbert space size" parameter into the RNN layer, analogous to rbm_multival.hpp.

**Comment by gcarleo** [comment]
Very nice generalisation @cmendl ! Maybe one can think of setting as a default value for the local size of the network the local size of the hilbert space? Anyway, @kchoo1118 has done some progress in understanding why AKLT is problematic. It seems that using the SR to optimise the energy is already much better, and the energy won't get stuck. However, for all networks on the AKLT it also seems that when approaching the ground state the acceptance of monte carlo moves goes to zero. Thus, most likely, there is an issue with the kind of moves we are doing in the sampling. For some reason they are not ergodic on approaching the ground state... 
Concerning the MPS layer, sure, that is certainly something worth trying. It is not entirely clear to me how to embed it in a general feed-forward network though (I guess an MPS layer only accepts discrete input values, right?)

**Comment by cmendl** [comment]
The last commit implements a RNN layer of MPS form. Formally the main difference to a vanilla RNN is that a rank-3 tensor is used to compute the next hidden state. In principle, the AKLT ground state should be recoverable exactly since it can be written as MPS with bond dimension 2 (using the identity function as activation function), except maybe up to boundary condition effects...

**Comment by cmendl** [comment]
Is there a built-in functionality of encoding the local states as unit vectors and feeding them into the "machine" as such? For example for spin 1/2 at each site, feeding spin-up as (1,0) and spin-down as (0,1) into the "machine", instead of 1 and -1?

**Comment by kchoo1118** [comment]
No I dont think so, would such a functionality be very useful?

On another note i managed to use your RNN to optimise for the AKLT state and the ising1d. For the AKLT somehow it was easier to optimise when one do the pi rotation about the z-axis on 1 sublattice (like Marshall sign rule). And for the ising1d, I find also that it is much more accurate with open boundary.

Also it seems better to use a sum output layer at the end, instead of a fully connected final layer.

**Comment by cmendl** [comment]
@kchoo1118 Thanks for the update! Open boundary conditions make sense due to the inherent sequential structure of RNNs.

Encoding local states as vectors is presumably the canonical approach for exact diagonalization for example, or the MPS layer class in the last commit. As proof of principle, I'll try to add an auxiliary layer to perform this encoding.

**Comment by gcarleo** [comment]
@cmendl , no, this functionality is not implemented. For ED we use a different approach, based on direct enumeration of the states. I guess you can add it as an ad-hoc encoding for the layers that need it? For example in RbmMultiVal  I used an ad hoc encoding to map the local values on binary entries

**Comment by cmendl** [comment]
@gcarleo The last commit implements an optional vectorization (similar to rbm_multival.hpp) in ffnn.hpp. I guess it makes more sense that a "machine" (instead of an individual layer) performs the vectorization since only the machine knows about the underlying Hilbert space.

@kchoo1118 and @everthemore I haven't tried it, but expect that the energy minimization for the AKLT state should work well now when using the MpsLayer and vectorizing the input as unit vectors of length 3 (hidden state dimension 2 should be sufficient).

**Comment by gcarleo** [comment]
@cmendl thanks for implementing this, it is nice to have the MPS layer working! 

The thing I believe should be rethought a little bit in this PR is how the vectorisation is handled. The problem is that the current approach is heavily intrusive for the FFNN class. Indeed the vectorisation is used only by a specific layer, whereas the change here is global for all the FFNN class and makes the code harder to read/maintain. 

I think it would be much more natural to include the vectorisation inside the layer itself, and not the machine.  Alternatively, isn't there a way to avoid completely this vectorisation step? 

In general, if the problem is that the layer doesn't know about the hilbert space, then this means that maybe the initial design of the layer class was not optimal, and we should modify the layers constructor to take a Hilbert object!

**Comment by cmendl** [comment]
Since the underlying Hilbert space is relevant as input to the first layer but not the downstream layers, it seems reasonable that the layer constructors generally don't receive a Hilbert object...

The modified FFNN class still defaults to no vectorization and behaves exactly as before in this case, but maybe one could (alternatively) code a separate "FFNNMultival" class for vectorization, analogous to RbmMultival?

**Comment by cmendl** [comment]
Implementing vectorisation as layer is not compatible with backpropagation I guess, i.e., it doesn't make sense to update an (integer) index by a continuous gradient... (On the other hand, this issue does not appear when vectorising the input to the network since its gradient is not required/used for updating the network parameters...)

**Comment by cmendl** [comment]
@kchoo1118 I've tried to run a ground state computation for the AKLT model using the MPS layer, but the computation did not finish within several hours. Is there a canonical way to track the progress of the simulation, or get some diagnosis if it's stuck somewhere, for example?

**Comment by kchoo1118** [comment]
The progress should be recorded in a .log file.
If the file is empty, there could be a bug in the code version (this was discovered a couple of weeks ago).
The issue was resolved in the latest versions. Maybe you could try merging the latest version on the master branch.

**Comment by cmendl** [comment]
@kchoo1118  Thx, the log file seems to work fine!

I've noticed that the MPS-type RNN layer doesn't work as well for the AKLT model as expected, so I'm currently revising it: the idea is to use a hidden state matrix (instead of a vector), starting from the identity matrix. The output of the MPS layer is then the last hidden state matrix (instead of the sequence of hidden states at each time step).

BTW the Hamiltonian constructed in Tutorials/AKLT/ now uses a convention where the analytically exact (and non-degenerate) lowest eigenvalue is 0. Thus it's easy to benchmark the numerical simulations... The ED version of the tutorial indeed reports a numerical zero, as expected.

**Comment by cmendl** [comment]
@gcarleo Is there a canonical way to output (or save) the network parameters after optimization? That would be useful for debugging.. Thx!

**Comment by kchoo1118** [comment]
The network parameters should be saved in the .wf file

**Comment by kchoo1118** [comment]
Oh sorry I thought you meant the wavefunction. Yes I agree with you I think that would be helpful.

**Comment by cmendl** [comment]
Great, thanks!

**Comment by gcarleo** [comment]
@cmendl  https://arxiv.org/pdf/1810.02352.pdf

**Comment by cmendl** [comment]
@gcarleo Thanks for pointing this out! I guess this renders the MPS layer obsolete for now...?

**Comment by cmendl** [comment]
@gcarleo Ok, thanks, I'll switch to implementing the unary RBM then...
